halo2_backend/src/plonk/circuit.rs

```diff
@@ -0,0 +1,390 @@
+
+        // Add an additional blinding factor as a slight defense against
+        // off-by-one errors.
+        factors + 1
+    }
+
+    /// Returns the minimum necessary rows that need to exist in order to
+    /// account for e.g. blinding factors.
+    pub fn minimum_rows(&self) -> usize {
+        self.blinding_factors() // m blinding factors
+            + 1 // for l_{-(m + 1)} (l_last)
+            + 1 // for l_0 (just for extra breathing room for the permutation
+                // argument, to essentially force a separation in the
+                // permutation polynomial between the roles of l_last, l_0
+                // and the interstitial values.)
+            + 1 // for at least one row
+    }
+
+    pub fn get_any_query_index(&self, column: ColumnMid, at: Rotation) -> usize {
+        let queries = match column.column_type {
+            Any::Advice(_) => &self.advice_queries,
+            Any::Fixed => &self.fixed_queries,
+            Any::Instance => &self.instance_queries,
+        };
+        for (index, instance_query) in queries.iter().enumerate() {
+            if instance_query == &(column, at) {
+                return index;
+            }
+        }
+        panic!("get_any_query_index called for non-existent query");
+    }
+
+    /// Returns the list of phases
+    pub fn phases(&self) -> impl Iterator<Item = u8> {
+        let max_phase = self
+            .advice_column_phase
+            .iter()
+            .max()
+            .copied()
+            .unwrap_or_default();
+        0..=max_phase
+    }
+
+    /// Obtain a pinned version of this constraint system; a structure with the
+    /// minimal parameters needed to determine the rest of the constraint
+    /// system.
+    pub fn pinned(&self) -> PinnedConstraintSystem<'_, F> {
+        PinnedConstraintSystem {
+            num_fixed_columns: &self.num_fixed_columns,
+            num_advice_columns: &self.num_advice_columns,
+            num_instance_columns: &self.num_instance_columns,
+            num_challenges: &self.num_challenges,
+            advice_column_phase: &self.advice_column_phase,
+            challenge_phase: &self.challenge_phase,
+            gates: PinnedGates(&self.gates),
+            fixed_queries: &self.fixed_queries,
+            advice_queries: &self.advice_queries,
+            instance_queries: &self.instance_queries,
+            permutation: &self.permutation,
+            lookups: &self.lookups,
+            shuffles: &self.shuffles,
+            minimum_degree: &self.minimum_degree,
+        }
+    }
+}
+
+struct PinnedGates<'a, F: Field>(&'a Vec<GateBack<F>>);
+
+impl<'a, F: Field> std::fmt::Debug for PinnedGates<'a, F> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> Result<(), std::fmt::Error> {
+        f.debug_list()
+            .entries(self.0.iter().map(|gate| &gate.poly))
+            .finish()
+    }
+}
+
+/// Represents the minimal parameters that determine a `ConstraintSystem`.
+#[allow(dead_code)]
```

Comment:
Maybe this can just be removed... It was inherited from before the frontend-backend split: https://github.com/privacy-scaling-explorations/halo2/blob/v0.3.0/halo2_proofs/src/plonk/circuit.rs#L1602
But the `Debug` implementation reads all the fields, so may Rust will not complain?

Reply:
Removed in https://github.com/privacy-scaling-explorations/halo2/pull/290/commits/e9f8c296deef613fbc3ae8e0c3662c67ff1468d6